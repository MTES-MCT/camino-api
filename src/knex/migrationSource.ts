import fs from 'fs'
import path from 'path'
import { Knex } from 'knex'

type MigrationFileExtension = typeof MIGRATION_FILE_EXTENSIONS[number]
const MIGRATION_FILE_EXTENSIONS = ['.js', '.ts', '.sql'] as const

interface MigrationFile {
  path: string
  extension: MigrationFileExtension
}
// Overloads fs-migration.js into Knex
// https://knexjs.org/#custom-migration-sources
export class CustomMigrationSource {
  private migrationFolders: readonly string[]

  constructor(migrationFolders: undefined | string | readonly string[]) {
    if (migrationFolders === undefined) {
      throw new Error('No migration folder configured')
    }
    if (typeof migrationFolders === 'string') {
      this.migrationFolders = [migrationFolders]
    } else {
      this.migrationFolders = migrationFolders
    }
  }

  public async getMigrations(): Promise<string[]> {
    const files = this.migrationFolders.flatMap(folder =>
      fs.readdirSync(folder)
    )

    return files
      .sort((a, b) => a.localeCompare(b))
      .filter(migration => {
        const migrationName = this.getMigrationName(migration)
        const extension = path.extname(migrationName)

        return (MIGRATION_FILE_EXTENSIONS as ReadonlyArray<string>).includes(
          extension
        )
      })
      .map(CustomMigrationSource.removeExtension)
  }

  private static removeExtension(migrationName: string): string {
    return migrationName.replace(/\.[^/.]+$/, '')
  }

  public getMigrationName(migration: string): string {
    return migration
  }

  private findMigrationFile(migration: string): MigrationFile | null {
    for (const migrationFileExtension of MIGRATION_FILE_EXTENSIONS) {
      for (const migrationFolder of this.migrationFolders) {
        const sqlMigrationFile = path.join(
          migrationFolder,
          `${migration}${migrationFileExtension}`
        )
        if (fs.existsSync(sqlMigrationFile)) {
          return {
            path: sqlMigrationFile,
            extension: migrationFileExtension
          }
        }
      }
    }

    return null
  }

  public getMigration(migration: string): {
    up: (knex: Knex) => Promise<void>
    down: () => Promise<void>
  } {
    const migrationFile = this.findMigrationFile(migration)
    if (migrationFile === null) {
      throw new Error(`Migration ${migration} not found`)
    }
    if (migrationFile.extension === '.sql') {
      return {
        up: async (knex: Knex): Promise<void> => {
          const sql = fs.readFileSync(migrationFile.path, 'utf8')
          await knex.raw(sql.replace(/\?/g, '\\?'))
        },
        down: async (): Promise<void> => {
          throw new Error(
            'Down migration not supported, restore backup instead'
          )
        }
      }
    } else {
      return require(migrationFile.path)
    }
  }
}
