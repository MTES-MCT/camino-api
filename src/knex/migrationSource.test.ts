import { CustomMigrationSource } from './migrationSource'
import { join } from 'path'

const migrationPath = join(__dirname, '/migrations-schema')

it('should return all the migrations', async () => {
  const tested = new CustomMigrationSource(migrationPath)
  const actual = await tested.getMigrations()

  // Testing only the 7 firsts which should never change
  expect(actual.slice(0, 7)).toMatchSnapshot()
})

it('should return a given migration', async () => {
  const tested = new CustomMigrationSource(migrationPath)
  const migrations = await tested.getMigrations()
  const firstMigration = await tested.getMigration(migrations[0])

  expect(firstMigration).not.toBeNull()
})
