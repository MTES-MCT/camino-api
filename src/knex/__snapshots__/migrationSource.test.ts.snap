// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return all the migrations 1`] = `
Array [
  "20180521000000_globales",
  "20180521000001_domaines_types_statuts",
  "20180521000002_demarches_etapes",
  "20180521000003_unites",
  "20180521000004_substances",
  "20180521000005_territoires",
  "20180521000006_calendrier",
]
`;
